torch/jit/python_frontend.py

```diff
@@ -0,0 +1,285 @@
+import torch
+import sys
+import ast
+import inspect
+from functools import partial
+from collections import namedtuple
+from torch._C._jit_tree_views import *
+
+# TODO: improve error reporting (show source)
+#       once this is done, convert all asserts into checks with nicer error messages
+
+PY2 = sys.version_info[0] == 2
+_reserved_prefix = '__jit'
+
+
+def get_jit_ast(fn):
+    source = dedent(inspect.getsource(fn))
+    py_ast = ast.parse(source)
+    if len(py_ast.body) != 1 or not isinstance(py_ast.body[0], ast.FunctionDef):
+        raise RuntimeError("expected a single top-level function")
+    return build_def(SourceRangeFactory(source), py_ast.body[0])
+
+
+def dedent(source):
+    lines = source.split('\n')
+    indent_depth = 0
+    while lines[0][indent_depth] == ' ':
+        indent_depth += 1
+    return '\n'.join(l[indent_depth:] for l in lines)
+
+
+class Builder(object):
```

Comment:
What is the need for this? Can't we just have a single object with `build_*` methods on it? Using that we would no longer have to pass `ctx` around everywhere.

Reply:
Closures are poor man's objects, objects are poor man's closures :wink: It's not like having them all in a single object would make the `ctx` disappear. You'd still have to take an extra `self` argument in the methods, and change `build_expr(ctx, ...)` to `self.build_expr(...)`, so it's not a win at all. I did it this way because:
1. This lets me separate handlers for different production types in different "scopes" (basically use the class definitions to group them), and it keeps the code more modular than if it was a single monolithic object for handling anything. Then, I only use this base class to have a single place where I will implement the nice "not supported" error, and they will both benefit from it. The interface seems quite simple too. I can add docstrings if you'd like.
1. I like the functional syntax more